klippy/extras/filament_switch_sensor.py

```diff
@@ -0,0 +1,165 @@
+# Generic Filament Sensor Module
+#
+# Copyright (C) 2019  Eric Callahan <arksine.code@gmail.com>
+#
+# This file may be distributed under the terms of the GNU GPLv3 license.
+import logging
+
+class BaseSensor(object):
+    def __init__(self, config):
+        self.name = config.get_name().split()[1]
+        self.printer = config.get_printer()
+        self.gcode = self.printer.lookup_object('gcode')
+        self.runout_gcode = config.get('runout_gcode', None)
+        self.autoload_gcode = config.get('autoload_gcode', None)
+        self.runout_pause = config.getboolean('pause_on_runout', True)
+        if self.runout_pause:
+            if self.runout_gcode is None:
+                self.runout_gcode = "PAUSE"
+            else:
+                self.runout_gcode = "PAUSE\n" + self.runout_gcode
+        self.runout_cb = self._runout_event_handler
+        self.insert_cb = self._autoload_event_handler
+        self.runout_enabled = False
+        self.insert_enabled = False
+        self.event_running = False
+        self.monitor_state = True
+        self.print_status = "idle"
+        for status in ["idle", "ready", "printing"]:
+            self.printer.register_event_handler(
+                "idle_timeout:%s" % (status),
+                (lambda e, s=self, st=status: s._update_print_status(st)))
+        self.printer.register_event_handler("klippy:ready", self._handle_ready)
+        self.gcode.register_mux_command(
+            "QUERY_FILAMENT_SENSOR", "SENSOR", self.name,
+            self.cmd_QUERY_FILAMENT_SENSOR,
+            desc=self.cmd_QUERY_FILAMENT_SENSOR_help)
+    def _handle_ready(self):
+        self.toolhead = self.printer.lookup_object('toolhead')
+        # delay turning on insert detection a bit.  This prevents an
+        # insert event triggering on startup
+        if self.autoload_gcode is not None:
+            reactor = self.printer.get_reactor()
+            reactor.register_callback(
+                (lambda e, s=self, st=self.print_status:
+                 s._update_print_status(st)),
+                reactor.monotonic() + 2.)
+    def _update_print_status(self, status):
+        logging.info(
+            "filament_sensor: print status changed to: %s" % (status))
```

Comment:
This adds logging for every printer state change.  I'm not sure that is desirable.  If it is desirable, it should be added to idle_timeout.py and be done separate from this series.

Reply:
You are correct, I used that to make sure the internal state was updating correctly, I will remove it.